File: source/Mach1DecodeEmscripten.cpp
#include <emscripten/bind.h>

#include "Mach1DecodeCAPI.h"
#include "include/cpp/Mach1Decode.cpp"

using namespace emscripten;

EMSCRIPTEN_BINDINGS(Mach1Decode) {
    enum_<Mach1PlatformType>("Mach1PlatformType")
        .value("Mach1PlatformDefault", Mach1PlatformDefault)
        .value("Mach1PlatformUnity", Mach1PlatformUnity)
        .value("Mach1PlatformUE", Mach1PlatformUE)
        .value("Mach1PlatformOfEasyCam", Mach1PlatformOfEasyCam)
        .value("Mach1PlatformAndroid", Mach1PlatformAndroid)
        .value("Mach1PlatformiOS", Mach1PlatformiOS)
        .value("Mach1PlatformiOSTableTop_ZVertical", Mach1PlatformiOSTableTop_ZVertical)
        .value("Mach1PlatformiOSPortraitHandheld_YVertical", Mach1PlatformiOSPortraitHandheld_YVertical)
        .value("Mach1PlatformiOSPortrait_YawOnly", Mach1PlatformiOSPortrait_YawOnly)
        ;
    
	enum_<Mach1DecodeAlgoType>("Mach1DecodeAlgoType")
        .value("Mach1DecodeAlgoSpatial_8", Mach1DecodeAlgoSpatial_8)
        .value("Mach1DecodeAlgoHorizon_4", Mach1DecodeAlgoHorizon_4)
        .value("Mach1DecodeAlgoHorizonPairs", Mach1DecodeAlgoHorizonPairs)
        .value("Mach1DecodeAlgoSpatial_12", Mach1DecodeAlgoSpatial_12)
        .value("Mach1DecodeAlgoSpatial_14", Mach1DecodeAlgoSpatial_14)
        .value("Mach1DecodeAlgoSpatial_18", Mach1DecodeAlgoSpatial_18)
        .value("Mach1DecodeAlgoSpatial_22", Mach1DecodeAlgoSpatial_22)
        .value("Mach1DecodeAlgoSpatial_32", Mach1DecodeAlgoSpatial_32)
        .value("Mach1DecodeAlgoSpatial_36", Mach1DecodeAlgoSpatial_36)
        .value("Mach1DecodeAlgoSpatial_48", Mach1DecodeAlgoSpatial_48)
        .value("Mach1DecodeAlgoSpatial_60", Mach1DecodeAlgoSpatial_60)
        ;

	register_vector<float>("VectorFloat");

    value_object<Mach1Point3D>("Mach1Point3D")
		.field("x", &Mach1Point3D::x)
		.field("y", &Mach1Point3D::y)
		.field("z", &Mach1Point3D::z)
        ;

    value_object<Mach1Point4D>("Mach1Point4D")
        .field("x", &Mach1Point4D::x)
        .field("y", &Mach1Point4D::y)
        .field("z", &Mach1Point4D::z)
        .field("w", &Mach1Point4D::w)
        ;

    class_<Mach1Decode>("Mach1DecodeInternal")
        .constructor<>()
        .function("setPlatformType", &Mach1Decode::setPlatformType)
        .function("setDecodeAlgoType", &Mach1Decode::setDecodeAlgoType)
        .function("decode", &Mach1Decode::decode)
        .function("decodeCoeffs", &Mach1Decode::decodeCoeffs)
        .function("setFilterSpeed", &Mach1Decode::setFilterSpeed)
        .function("getFormatChannelCount", &Mach1Decode::getFormatChannelCount)
        .function("setRotation", &Mach1Decode::setRotation)
        .function("setRotationDegrees", &Mach1Decode::setRotationDegrees)
        .function("setRotationRadians", &Mach1Decode::setRotationRadians)
        .function("setRotationQuat", &Mach1Decode::setRotationQuat)
        .function("beginBuffer", &Mach1Decode::beginBuffer)
        .function("endBuffer", &Mach1Decode::endBuffer)
        .function("getCurrentTime", &Mach1Decode::getCurrentTime)
        .function("getCurrentAngle", &Mach1Decode::getCurrentAngle)
        .function("getLog", &Mach1Decode::getLog)
        ;
}

File: source/Mach1DecodeEmscripten.js
Module['onRuntimeInitialized'] = function() {
    console.log('loaded Mach1Decode');

    Module['Mach1Decode'] = function() {
        this.Mach1PlatformType = {
            "Mach1PlatformDefault": (Module).Mach1PlatformType.Mach1PlatformDefault,
            "Mach1PlatformUnity": (Module).Mach1PlatformType.Mach1PlatformUnity,
            "Mach1PlatformUE": (Module).Mach1PlatformType.Mach1PlatformUE,
            "Mach1PlatformOfEasyCam": (Module).Mach1PlatformType.Mach1PlatformOfEasyCam,
            "Mach1PlatformAndroid": (Module).Mach1PlatformType.Mach1PlatformAndroid,
            "Mach1PlatformiOS": (Module).Mach1PlatformType.Mach1PlatformiOS,
            "Mach1PlatformiOSTableTop_ZVertical": (Module).Mach1PlatformiOSTableTop_ZVertical,
            "Mach1PlatformiOSPortraitHandheld_YVertical": (Module).Mach1PlatformiOSPortraitHandheld_YVertical,
            "Mach1PlatformiOSPortrait_YawOnly": (Module).Mach1PlatformiOSPortrait_YawOnly,
        };

        this.Mach1DecodeAlgoType = {
            "Mach1DecodeAlgoSpatial_8": (Module).Mach1DecodeAlgoType.Mach1DecodeAlgoSpatial_8,
            "Mach1DecodeAlgoHorizon_4": (Module).Mach1DecodeAlgoType.Mach1DecodeAlgoHorizon_4,
            "Mach1DecodeAlgoHorizonPairs": (Module).Mach1DecodeAlgoType.Mach1DecodeAlgoHorizonPairs,
            "Mach1DecodeAlgoSpatial_12": (Module).Mach1DecodeAlgoType.Mach1DecodeAlgoSpatial_12,
            "Mach1DecodeAlgoSpatial_14": (Module).Mach1DecodeAlgoType.Mach1DecodeAlgoSpatial_14,
            "Mach1DecodeAlgoSpatial_32": (Module).Mach1DecodeAlgoType.Mach1DecodeAlgoSpatial_32,
            "Mach1DecodeAlgoSpatial_36": (Module).Mach1DecodeAlgoType.Mach1DecodeAlgoSpatial_36,
            "Mach1DecodeAlgoSpatial_48": (Module).Mach1DecodeAlgoType.Mach1DecodeAlgoSpatial_48,
            "Mach1DecodeAlgoSpatial_60": (Module).Mach1DecodeAlgoType.Mach1DecodeAlgoSpatial_60,
        };

        this.delete = function() {
            if (this._m1obj) {
                this._m1obj.delete();
                this._m1obj = null;
            }
        };

        this.setPlatformType = function(platform) {
            if (this._m1obj) {
                this._m1obj.setPlatformType(platform);
            }
        };

        this.setDecodeAlgoType = function(decodeAlgo) {
            if (this._m1obj) {
                this._m1obj.setDecodeAlgoType(decodeAlgo);
            }
        };

        this.setFilterSpeed = function(filterSpeed) {
            if (this._m1obj) {
                this._m1obj.setFilterSpeed(filterSpeed);
            }
        };

        this.setRotation = function(yaw, pitch, roll) {
            if (this._m1obj) {
                this._m1obj.setRotation(yaw, pitch, roll);
            }
        };

        this.setRotationDegrees = function(yaw, pitch, roll) {
            if (this._m1obj) {
                this._m1obj.setRotationDegrees(yaw, pitch, roll);
            }
        };

        this.setRotationRadians = function(yaw, pitch, roll) {
            if (this._m1obj) {
                this._m1obj.setRotationRadians(yaw, pitch, roll);
            }
        };

        this.setRotationQuat = function(w, x, y, z) {
            if (this._m1obj) {
                this._m1obj.setRotationQuat(x, y, z, w);
            }
        };

        this.getFormatChannelCount = function() {
            if (this._m1obj) {
                return this._m1obj.getFormatChannelCount();
            }
        }

        this.getCurrentAngle = function() {
            if (this._m1obj) {
                return this._m1obj.getCurrentAngle();
            }
        };

        this.decode = function(yaw, pitch, roll, bufferSize, sampleIndex) {
            if (this._m1obj) {
                var v = this._m1obj.decode(yaw, pitch, roll, bufferSize ? bufferSize : 0, sampleIndex ? sampleIndex : 0);
                var a = [];
                for (var i = 0, c = v.size(); i < c; i++) {
                    a[i] = v.get(i);
                }
                v.delete();
                return a;
            }
        };

        this.decodeCoeffs = function(bufferSize, sampleIndex) {
            if (this._m1obj) {
                var v = this._m1obj.decodeCoeffs(bufferSize ? bufferSize : 0, sampleIndex ? sampleIndex : 0);
                var a = [];
                for (var i = 0, c = v.size(); i < c; i++) {
                    a[i] = v.get(i);
                }
                v.delete();
                return a;
            }
        };

        this.beginBuffer = function() {
            if (this._m1obj) {
                this._m1obj.beginBuffer();
            }
        };

        this.endBuffer = function() {
            if (this._m1obj) {
                this._m1obj.endBuffer();
            }
        };

        this._m1obj = new(Module).Mach1DecodeInternal();
    };
};